subprojects/core/src/main/java/org/gradle/configuration/BuildOperationScriptPlugin.java

```diff
@@ -43,12 +47,25 @@ public ScriptSource getSource() {
 
     @Override
     public void apply(final Object target) {
-        String operationDisplayNamePrefix = "Apply " + getSource().getDisplayName() + " to " + target;
-        buildOperationExecutor.run(operationDisplayNamePrefix, new Action<BuildOperationContext>() {
-            @Override
-            public void execute(BuildOperationContext buildOperationContext) {
-                decorated.apply(target);
-            }
-        });
+        TextResource resource = getSource().getResource();
+        if (resource.isContentCached() && resource.getHasEmptyContent()) {
+            //no operation, if there is no script code provided
+            decorated.apply(target);
+        } else {
+            buildOperationExecutor.run(computeBuildOperationDetails(target), new Action<BuildOperationContext>() {
+                @Override
+                public void execute(BuildOperationContext buildOperationContext) {
+                    decorated.apply(target);
+                }
+            });
+        }
+    }
+
+    private BuildOperationDetails computeBuildOperationDetails(Object target) {
+        ScriptSource source = getSource();
+        File file = source.getResource().getFile();
+        String name = "Apply script " + (file != null ? file.getName() : source.getDisplayName());
+        String displayName = name + " to " + target;
```

Comment:
We should not use `getSource().getDisplayName()` as it uses `getPath()` for files, which usually provides the absolute path.

`file.getName()` is a compromise. Better would be the project-relative path. But we do not have that here and do not want to do additional computation. It is usually clear from the `target`, which file it is (e.g. which `build.gradle` of which project). The only ambiguous case would be if several custom scripts with the same name, but from different folders, are applied in the same project. A case that probably does not happen much in practice. 

In general, the `ScriptSource` object is now attached as `operationDescriptor` allowing consumers to get more information about the source.